src/analyser.cpp

```diff
@@ -1115,12 +1248,835 @@ void Analyser::AnalyserImpl::analyseEquationAst(const AnalyserEquationAstPtr &as
+                }
+            }
+        }
+
+        for (const auto &unitsItem : unitsMap) {
+            if ((unitsItem.first != "dimensionless")
+                && !areNearlyEqual(unitsItem.second, 0.0)) {
+                auto intExponent = int(unitsItem.second);
+                auto exponent = areNearlyEqual(unitsItem.second, intExponent) ?
+                                    convertToString(intExponent) :
+                                    convertToString(unitsItem.second, false);
+
+                if (!unit.empty()) {
+                    unit += " x ";
+                }
+
+                unit += unitsItem.first;
+
+                if (exponent != "1") {
+                    unit += "^" + exponent;
+                }
+            }
+        }
+
+        if (!unit.empty()) {
+            units.push_back(unit);
+        }
+    }
+
+    std::string unitsString;
+
+    for (size_t i = 0; i < units.size(); ++i) {
+        if (i > 0) {
+            unitsString += (i == units.size() - 1) ? " and " : ", ";
+        }
+
+        unitsString += "'" + units[i] + "'";
+    }
+
+    return unitsString;
+}
+
+std::string Analyser::AnalyserImpl::expressionUnits(const AnalyserEquationAstPtr &ast,
+                                                    const UnitsMaps &unitsMaps,
+                                                    const UnitsMaps &userUnitsMaps,
+                                                    const UnitsMultipliers &unitsMultipliers)
+{
+    // Return a string version of the given AST and (user) units maps and units
+    // multipliers.
+
+    auto res = expression(ast, false) + " is ";
+    auto unitsString = expressionUnits(unitsMaps, unitsMultipliers);
+    auto userUnitsString = expressionUnits(userUnitsMaps);
+
+    if (userUnitsString.empty()) {
+        res += "'dimensionless'";
+    } else {
+        res += "in " + userUnitsString;
+
+        if (!unitsString.empty() && (unitsString != userUnitsString)) {
+            res += " (i.e. " + unitsString + ")";
+        }
+    }
+
+    return res;
+}
+
+void Analyser::AnalyserImpl::defaultUnitsMapsAndMultipliers(UnitsMaps &unitsMaps,
+                                                            UnitsMaps &userUnitsMaps,
+                                                            UnitsMultipliers &unitsMultipliers)
+{
+    // Default units maps and multipliers.
+
+    unitsMaps = {UnitsMap()};
+    userUnitsMaps = {UnitsMap()};
+    unitsMultipliers = {0.0};
+}
+
+void Analyser::AnalyserImpl::analyseEquationUnits(const AnalyserEquationAstPtr &ast,
+                                                  UnitsMaps &unitsMaps,
+                                                  UnitsMaps &userUnitsMaps,
+                                                  UnitsMultipliers &unitsMultipliers,
+                                                  Strings &issueDescriptions)
+{
+    // Analyse the units used with different MathML elements (table 2.1 of the
```

Comment:
note to andre: make an issue to pull this information out into the documentation at some point...

Reply:
see #834